dask_match/io/tests/test_io.py

```diff
@@ -111,6 +112,39 @@ def test_predicate_pushdown(tmpdir):
     assert all(y_result["b"] == 4)
 
 
+@pytest.mark.parametrize("fmt", ["parquet", "csv", "pandas"])
+def test_io_culling(tmpdir, fmt):
+    pdf = pd.DataFrame({c: range(10) for c in "abcde"})
+    if fmt == "parquet":
+        dd.from_pandas(pdf, 2).to_parquet(tmpdir)
+        df = read_parquet(tmpdir)
+    elif fmt == "parquet":
+        dd.from_pandas(pdf, 2).to_csv(tmpdir)
+        df = read_csv(tmpdir + "/*")
+    else:
+        df = from_pandas(pdf, 2)
+    df = (df[["a", "b"]] + 1).partitions[1]
+    df2 = optimize(df)
+
+    # All tasks should be fused for the single output partition
+    assert df2.npartitions == 1
+    assert len(df2.dask) == df2.npartitions
+    expected = pdf.iloc[5:][["a", "b"]] + 1
+    assert_eq(df2, expected, check_index=False)
+
+    def _check_culling(expr, partitions):
+        """CHeck that _partitions is set to the expected value"""
+        for dep in expr.dependencies():
+            _check_culling(dep, partitions)
+        if "_partitions" in expr._parameters:
+            assert expr._partitions == partitions
```

Comment:
I moved this out of `utils.py`.  I understand that this means that we can't test datasets.timeseries.  I think that in the future we should pull out `assert_eq` and make our own.  This would then go there and be more universal.  Until then I'd like to have this isolated-ish if you're cool with it.

Reply:
Works for me